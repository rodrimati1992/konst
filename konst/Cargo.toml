[package]
name = "konst"
version = "0.3.5"
authors = ["rodrimati1992 <rodrimatt1985@gmail.com>"]
rust-version = "1.65.0"
edition = "2021"
license = "Zlib"
description = "Const equivalents of std functions, compile-time comparison, and parsing"
documentation = "https://docs.rs/konst/"
readme="../README.md"
keywords = ["no-std", "const_fn", "parsing"]
categories = ["no-std", "parsing"]
repository = "https://github.com/rodrimati1992/konst/"
include = [
    "Cargo.toml", 
    "src/**/*.rs", 
    "../README.md",
    "LICENSE-ZLIB.md", 
]


[dependencies.konst_proc_macros]
version = "=0.3.0"
path = "../konst_proc_macros"
optional = true

[dependencies.konst_kernel]
version = "=0.3.5"
path = "../konst_kernel"
features = ["rust_1_64", "__for_konst"]

[dev-dependencies.rand]
version = "0.8.4"
default-features = false
features = ["small_rng"]

# dev-dependencies can't be optional
[dependencies.trybuild]
version = "1.0"
optional = true

[dependencies.typewit]
version = "1.1"
features = ["rust_1_61"]

[dependencies.const_panic]
version = "0.2.6"
features = ["rust_1_64"]
default-features = false

[features]
default = ["cmp", "iter", "parsing_proc"]


# Enables extra checks for debug
debug = ["konst_kernel/debug"]
cmp = []
iter = ["konst_kernel/iter"]
parsing_proc = ["parsing", "konst_proc_macros"]
parsing = []
rust_latest_stable = []
mut_refs = ["rust_latest_stable", "konst_kernel/mut_refs"]
nightly_mut_refs = ["mut_refs", "konst_kernel/nightly_mut_refs"]

alloc = ["konst_kernel/alloc"]

# Enables ui tests, which are intended for the latest stable release.
# This is not enabled in CI, because it's way too sensitive to 
# changes in how rustc formats errors
__ui = ["trybuild", "rust_latest_stable"]
__cp_derive = ["const_panic/derive", "const_panic/non_basic"]

docsrs = ["konst_kernel/docsrs"]



[package.metadata.docs.rs]
features = ["docsrs", "alloc", "cmp", "parsing_proc", "rust_latest_stable", "nightly_mut_refs"]

