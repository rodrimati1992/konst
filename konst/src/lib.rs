//! Const equivalents of std functions and const parsing.
//!
//! # Features
//!
//! This crate provides:
//!
//! - Const fn equivalents of standard library functions and methods.
//!
//! - Compile-time parsing through the [`Parser`] type, and [`parser_method`] macro.
//!
//! # Examples
//!
//! ### Parsing an enum
//!
//! This example demonstrates how you can parse a simple enum from an environment variable,
//! at compile-time.
//!
//! ```rust
//! use konst::{
//!     eq_str,
//!     option,
//!     result::unwrap_ctx,
//! };
//!
//! #[derive(Debug, PartialEq)]
//! enum Direction {
//!     Forward,
//!     Backward,
//!     Left,
//!     Right,
//! }
//!
//! impl Direction {
//!     const fn try_parse(input: &str) -> Result<Self, ParseDirectionError> {
//!         // As of Rust 1.65.0, string patterns don't work in const contexts
//!         match () {
//!             _ if eq_str(input, "forward") => Ok(Direction::Forward),
//!             _ if eq_str(input, "backward") => Ok(Direction::Backward),
//!             _ if eq_str(input, "left") => Ok(Direction::Left),
//!             _ if eq_str(input, "right") => Ok(Direction::Right),
//!             _ => Err(ParseDirectionError),
//!         }
//!     }
//! }
//!
//! const CHOICE: &str = option::unwrap_or!(option_env!("chosen-direction"), "forward");
//!
//! const DIRECTION: Direction = unwrap_ctx!(Direction::try_parse(CHOICE));
//!
//! fn main() {
//!     match DIRECTION {
//!         Direction::Forward => assert_eq!(CHOICE, "forward"),
//!         Direction::Backward => assert_eq!(CHOICE, "backward"),
//!         Direction::Left => assert_eq!(CHOICE, "left"),
//!         Direction::Right => assert_eq!(CHOICE, "right"),
//!     }
//! }
//!
//! #[derive(Debug, PartialEq)]
//! pub struct ParseDirectionError;
//!
//! use std::fmt::{self, Display};
//!
//! impl Display for ParseDirectionError {
//!     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
//!         f.write_str("Failed to parse a Direction")
//!     }
//! }
//!
//! impl ParseDirectionError {
//!     const fn panic(&self) -> ! {
//!         panic!("failed to parse a Direction")
//!     }
//! }
//!
//! ```
//!
//! ### Parsing CSV
//!
//! This example demonstrates how CSV can be parsed into integers.
//!
//! This example requires the `"parsing"` and `"iter"` features
//! (both are enabled by default).
//!
#![cfg_attr(all(feature = "parsing", feature = "iter"), doc = "```rust")]
#![cfg_attr(not(all(feature = "parsing", feature = "iter")), doc = "```ignore")]
//! use konst::{
//!     primitive::parse_u64,
//!     result::unwrap_ctx,
//!     iter, string,
//! };
//!
//! const CSV: &str = "3, 8, 13, 21, 34";
//!
//! static PARSED: [u64; 5] = iter::collect_const!(u64 =>
//!     string::split(CSV, ","),
//!         map(string::trim),
//!         map(|s| unwrap_ctx!(parse_u64(s))),
//! );
//!
//! assert_eq!(PARSED, [3, 8, 13, 21, 34]);
//!
//! ```
//!
//! ### Parsing a struct
//!
//! This example demonstrates how a key-value pair format can be parsed into a struct.
//!
//! This requires the `"parsing_proc"` feature (enabled by default).
//!
#![cfg_attr(feature = "parsing_proc", doc = "```rust")]
#![cfg_attr(not(feature = "parsing_proc"), doc = "```ignore")]
//! use konst::{
//!     parsing::{Parser, ParseValueResult},
//!     eq_str,
//!     for_range, parser_method, try_, unwrap_ctx,
//! };
//!
//! const PARSED: Struct = {
//!     // You can also parse strings from environment variables, or from an `include_str!(....)`
//!     let input = "\
//!         colors = red, blue, green, blue
//!         amount = 1000
//!         repeating = circle
//!         name = bob smith
//!     ";
//!     
//!     unwrap_ctx!(parse_struct(Parser::new(input))).0
//! };
//!
//! fn main(){
//!     assert_eq!(
//!         PARSED,
//!         Struct{
//!             name: "bob smith",
//!             amount: 1000,
//!             repeating: Shape::Circle,
//!             colors: [Color::Red, Color::Blue, Color::Green, Color::Blue],
//!         }
//!     );
//! }
//!
//! #[derive(Debug, Clone, PartialEq, Eq)]
//! pub struct Struct<'a> {
//!     pub name: &'a str,
//!     pub amount: usize,
//!     pub repeating: Shape,
//!     pub colors: [Color; 4],
//! }
//!
//! #[derive(Debug, Clone, PartialEq, Eq)]
//! pub enum Shape {
//!     Circle,
//!     Square,
//!     Line,
//! }
//!
//! #[derive(Debug, Copy, Clone, PartialEq, Eq)]
//! pub enum Color {
//!     Red,
//!     Blue,
//!     Green,
//! }
//!
//! pub const fn parse_struct(mut parser: Parser<'_>) -> ParseValueResult<'_, Struct<'_>> {
//!     let mut name = "<none>";
//!     let mut amount = 0;
//!     let mut repeating = Shape::Circle;
//!     let mut colors = [Color::Red; 4];
//!     
//!     parser = parser.trim_end();
//!     if !parser.is_empty() {
//!         loop {
//!             let mut prev_parser = parser.trim_start();
//!
//!             parser = try_!(parser.find_skip('='));
//!
//!             parser_method!{prev_parser, strip_prefix;
//!                 "name" => (name, parser) = try_!(parser.trim_start().split_keep('\n')),
//!                 "amount" => (amount, parser) = try_!(parser.trim_start().parse_usize()),
//!                 "repeating" => (repeating, parser) = try_!(parse_shape(parser.trim_start())),
//!                 "colors" => (colors, parser) = try_!(parse_colors(parser.trim_start())),
//!                 _ => {
//!                     let err = &"could not parse Struct field name";
//!                     return Err(prev_parser.into_other_error(err));
//!                 }
//!             }
//!
//!             if parser.is_empty() {
//!                 break
//!             }
//!             parser = try_!(parser.strip_prefix("\n"));
//!         }
//!     }
//!
//!     Ok((Struct{name, amount, repeating, colors}, parser))
//! }
//!
//! pub const fn parse_shape(mut parser: Parser<'_>) -> ParseValueResult<'_, Shape> {
//!     let shape = parser_method!{parser, strip_prefix;
//!         "circle" => Shape::Circle,
//!         "square" => Shape::Square,
//!         "line" => Shape::Line,
//!         _ => return Err(parser.into_other_error(&"could not parse Shape"))
//!     };
//!     Ok((shape, parser))
//! }
//!
//! pub const fn parse_colors<const LEN: usize>(
//!     mut parser: Parser<'_>,
//! ) -> ParseValueResult<'_, [Color; LEN]> {
//!     let mut colors = [Color::Red; LEN];
//!
//!     for_range!{i in 0..LEN =>
//!         (colors[i], parser) = try_!(parse_color(parser.trim_start()));
//!         
//!         match parser.strip_prefix(",") {
//!             Ok(next) => parser = next,
//!             Err(_) if i == LEN - 1 => {}
//!             Err(e) => return Err(e),
//!         }
//!     }
//!
//!     Ok((colors, parser))
//! }
//!
//! pub const fn parse_color(mut parser: Parser<'_>) -> ParseValueResult<'_, Color> {
//!     let color = parser_method!{parser, strip_prefix;
//!         "red" => Color::Red,
//!         "blue" => Color::Blue,
//!         "green" => Color::Green,
//!         _ => return Err(parser.into_other_error(&"could not parse Color"))
//!     };
//!     Ok((color, parser))
//! }
//!
//!
//!
//! ```
//!
//! # Cargo features
//!
//! These are the features of these crates:
//!
//! - `"iter"`(enabled by default):
//! Enables all iteration items, including macros/functions that take/return iterators,
//!
//! - `"cmp"`(enabled by default):
//! Enables all comparison functions and macros,
//! the string equality and ordering comparison functions don't require this feature.
//!
//! - `"parsing_proc"`(enabled by default):
//! Enables the `"parsing"` feature, compiles the `konst_proc_macros` dependency,
//! and enables the [`parser_method`] macro.
//! You can use this feature instead of `"parsing"` if the slightly longer
//! compile times aren't a problem.
//!
//! - `"parsing"`(enabled by default):
//! Enables the [`parsing`] module (for parsing from `&str` and `&[u8]`),
//! the `primitive::parse_*` functions, `try_rebind`, and `rebind_if_ok` macros.
//!
//! - `"alloc"`:
//! Enables items that use types from the [`alloc`] crate, including `Vec` and `String`.
//!
//! ### Rust release related
//!
//! None of thse features are enabled by default.
//!
//! - `"rust_latest_stable"`: enables the latest `"rust_1_*"` feature(there's currently none).
//! Only recommendable if you can update the Rust compiler every stable release.
//!
//! - `"mut_refs"`(disabled by default):
//! Enables const functions that take mutable references.
//! Use this whenever mutable references in const contexts are stabilized.
//! Also enables the `"rust_latest_stable"` feature.
//!
//! - `"nightly_mut_refs"`(disabled by default):
//! Enables the `"mut_refs"` feature. Requires Rust nightly.
//!
//! # No-std support
//!
//! `konst` is `#![no_std]`, it can be used anywhere Rust can be used.
//!
//! # Minimum Supported Rust Version
//!
//! `konst` requires Rust 1.65.0.
//!
//! Features that require newer versions of Rust, or the nightly compiler,
//! need to be explicitly enabled with crate features.
//!
//!
//!
//! [`alloc`]: https://doc.rust-lang.org/alloc/
//! [`const_eq`]: ./macro.const_eq.html
//! [`const_eq_for`]: ./macro.const_eq_for.html
//! [`const_cmp`]: ./macro.const_cmp.html
//! [`const_cmp_for`]: ./macro.const_cmp_for.html
//! [`parsing`]: ./parsing/index.html
//! [`primitive`]: ./primitive/index.html
//! [`parser_method`]: macro.parser_method.html
//! [`Parser`]: ./parsing/struct.Parser.html
//! [`Parser::parse_u128`]: ./parsing/struct.Parser.html#method.parse_u128
//!
#![deny(missing_docs)]
#![deny(unused_results)]
// clippy's opinionated BS
#![allow(clippy::needless_doctest_main)]
#![allow(clippy::init_numbered_fields)]
////////////
#![forbid(clippy::missing_const_for_fn)]
#![cfg_attr(
    feature = "nightly_mut_refs",
    feature(const_mut_refs, const_slice_from_raw_parts_mut)
)]
#![cfg_attr(feature = "docsrs", feature(doc_cfg))]
#![no_std]

#[cfg(feature = "alloc")]
extern crate alloc;

include! {"./root_module_macros.rs"}

#[macro_use]
mod macros;

#[doc(hidden)]
pub mod __for_cmp_impls;

// pub mod other;

#[cfg(feature = "alloc")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "alloc")))]
pub mod alloc_type;

pub mod array;

pub mod chr;

#[cfg(feature = "cmp")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "cmp")))]
pub mod cmp;

#[cfg(feature = "iter")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "iter")))]
pub mod iter;

/// const equivalents of `core::ffi` functions
pub mod ffi;

pub mod polymorphism;

pub mod primitive;

pub mod option;

pub mod result;

pub mod range;

pub mod maybe_uninit;

pub mod manually_drop;

pub mod nonzero;

pub mod other;

#[cfg(feature = "parsing")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "parsing")))]
pub mod parsing;

pub mod ptr;

mod utils;

#[allow(unused_imports)]
mod utils_1_56 {
    pub(crate) use konst_kernel::__priv_transmute;
}

#[cfg(feature = "parsing")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "parsing")))]
pub use crate::parsing::Parser;

#[cfg(feature = "parsing_proc")]
#[doc(hidden)]
pub use konst_proc_macros::{__priv_bstr_end, __priv_bstr_start};

pub mod slice;

pub mod string;

pub use ::const_panic;

pub use crate::string::{cmp_str, eq_str};

#[doc(no_inline)]
pub use crate::result::unwrap_ctx;

#[cfg(feature = "cmp")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "cmp")))]
pub use crate::string::{cmp_option_str, eq_option_str};

#[cfg(all(doctest, feature = "iter", feature = "parsing_proc"))]
#[doc = include_str!("../../README.md")]
pub struct ReadmeTest;

#[doc(hidden)]
pub mod __ {
    pub use core::{
        cmp::Ordering::{self, Equal, Greater, Less},
        compile_error, matches,
        ops::Range,
        option::Option::{self, None, Some},
        result::Result::{self, Err, Ok},
    };

    pub mod v {
        pub use core::{
            option::Option::Some,
            result::Result::{Err, Ok},
        };
    }

    pub use crate::__for_cmp_impls::U8Ordering;

    #[cfg(feature = "cmp")]
    #[cfg_attr(feature = "docsrs", doc(cfg(feature = "cmp")))]
    pub use crate::cmp::{CmpWrapper, ConstCmp, IsAConstCmp, IsNotStdKind, IsStdKind};

    pub use const_panic::concat_panic;
}
