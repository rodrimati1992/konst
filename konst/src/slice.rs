//! `const fn` equivalents of slice methods.
//!
//! # Removed in 0.4.0
//!
//! These functions were removed in 0.4.0 because there is an equivalent
//! const fn in the standard library:
//!
//! - `as_chunks`: [slice::as_chunks]
//! - `as_rchunks`: [slice::as_rchunks]
//! - `as_chunks_mut`: [slice::as_chunks_mut]
//! - `as_rchunks_mut`: [slice::as_rchunks_mut]
//! - `bytes_trim_end`: [slice::trim_ascii_end]
//! - `bytes_trim_start`: [slice::trim_ascii_start]
//! - `bytes_trim`: [slice::trim_ascii]
//! - `first_mut`: [slice::first_mut]
//! - `last_mut`: [slice::last_mut]
//! - `split_first_mut`: [slice::split_first_mut]
//! - `split_last_mut`: [slice::split_last_mut]
//!
//! The `array_chunks*` functions were removed in 0.4.0 because the
//! unstable equivalent function was removed.
//! You can replace instances of `konst::slice::array_chunks(slice)`
//! with `konst::slice::iter(slice.as_chunks().0)`.
//!
//!
//!

use core::fmt::{self, Display};

/// `const fn`s for comparing slices for equality and ordering.
#[cfg(feature = "cmp")]
#[cfg_attr(feature = "docsrs", doc(cfg(feature = "cmp")))]
pub mod cmp;

mod bytes_pattern;
mod slice_concatenation;
mod slice_const_methods;
mod slice_filler;

#[cfg(feature = "iter")]
mod slice_iter_methods;

pub use bytes_pattern::BytesPattern;

pub(crate) use bytes_pattern::PatternNorm;

pub use self::slice_concatenation::*;
pub use self::slice_const_methods::*;
pub use self::slice_filler::*;

#[cfg(feature = "iter")]
pub use slice_iter_methods::*;

__declare_slice_cmp_fns! {
    import_path = "konst",

    (
        ///
        ///  # Example
        ///
        ,
        /// ```rust
        /// use konst::slice::eq_bytes;
        ///
        /// const FOO: &[u8] = b"foo";
        /// const BAR: &[u8] = b"fooooo";
        /// const BAZ: &[u8] = b"bar";
        ///
        ///
        /// const FOO_EQ_FOO: bool = eq_bytes(FOO, FOO);
        /// assert!( FOO_EQ_FOO );
        ///
        /// const FOO_EQ_BAR: bool = eq_bytes(FOO, BAR);
        /// assert!( !FOO_EQ_BAR );
        ///
        /// const FOO_EQ_BAZ: bool = eq_bytes(FOO, BAZ);
        /// assert!( !FOO_EQ_BAZ );
        ///
        /// ```
        ///
        ,
        /// ```rust
        /// use konst::slice::cmp_bytes;
        ///
        /// use std::cmp::Ordering;
        ///
        /// const FOO: &[u8] = b"foo";
        /// const BAR: &[u8] = b"fooooo";
        /// const BAZ: &[u8] = b"bar";
        ///
        ///
        /// const FOO_CMP_FOO: Ordering = cmp_bytes(FOO, FOO);
        /// assert_eq!(FOO_CMP_FOO, Ordering::Equal);
        ///
        /// const FOO_CMP_BAR: Ordering = cmp_bytes(FOO, BAR);
        /// assert_eq!(FOO_CMP_BAR, Ordering::Less);
        ///
        /// const FOO_CMP_BAZ: Ordering = cmp_bytes(FOO, BAZ);
        /// assert_eq!(FOO_CMP_BAZ, Ordering::Greater);
        ///
        /// ```
        ///
        ,
        u8,
        eq_bytes,
        cmp_bytes,
    )
}

__declare_fns_with_docs! {
    (Option<&'a [u8]>, (eq_option_bytes, cmp_option_bytes))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a,]
        params(l, r)
        eq_comparison = eq_bytes(l, r),
        cmp_comparison = cmp_bytes(l, r),
        parameter_copyability = copy,
    ),
}

////////////////////////////////////////////////////////////////////////////////

/// The error produced by trying to convert from
/// `&[T]` to `&[T; N]`, or from `&mut [T]` to `&mut [T; N]`.
#[derive(Debug, PartialEq, Eq, Copy, Clone)]
pub struct TryIntoArrayError {
    slice_len: usize,
    array_len: usize,
}

impl Display for TryIntoArrayError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "could not convert slice of length `{}` to array of length `{}`",
            self.slice_len, self.array_len,
        )
    }
}

const _: () = {
    use const_panic::{
        PanicFmt, PanicVal, flatten_panicvals,
        fmt::{self as cfmt, ComputePvCount, FmtArg, FmtKind},
    };

    impl PanicFmt for TryIntoArrayError {
        type This = Self;
        type Kind = const_panic::IsCustomType;

        const PV_COUNT: usize = ComputePvCount {
            field_amount: 2,
            summed_pv_count: <usize>::PV_COUNT * 2,
            delimiter: cfmt::TypeDelim::Braced,
        }
        .call();
    }

    impl TryIntoArrayError {
        /// Formats a TryIntoArrayError
        pub const fn to_panicvals(
            &self,
            fmtarg: FmtArg,
        ) -> [PanicVal<'static>; TryIntoArrayError::PV_COUNT] {
            match fmtarg.fmt_kind {
                FmtKind::Debug => {
                    flatten_panicvals! {fmtarg;
                        "TryIntoArrayError",
                        open: cfmt::OpenBrace,
                            "slice_len: ", usize => self.slice_len, cfmt::COMMA_SEP,
                            "array_len: ", usize => self.array_len, cfmt::COMMA_TERM,
                        close: cfmt::CloseBrace,
                    }
                }
                _ => const_panic::utils::flatten_panicvals(&[&[
                    PanicVal::write_str("could not convert slice of length `"),
                    PanicVal::from_usize(self.slice_len, FmtArg::DEBUG),
                    PanicVal::write_str("` to array of length `"),
                    PanicVal::from_usize(self.array_len, FmtArg::DEBUG),
                    PanicVal::write_str("`"),
                ]]),
            }
        }
    }
};

////////////////////////////////////////////////////////////////////////////////

/// Tries to convert from `&[T]` to `&[T; N]`.
///
/// Returns an `Err(TryIntoArrayError{..})` when the slice doesn't match the expected length.
///
/// # Example
///
/// ```rust
/// use konst::{
///     slice::{TryIntoArrayError, try_into_array},
///     result,
/// };
///
///
/// const fn arr_5() -> Option<&'static [u64; 5]> {
///     let slice: &[u64] = &[1, 10, 100, 1000, 10000];
///
///     // Passing the length explicitly to the function
///     result::ok!(try_into_array::<_, 5>(slice))
/// }
///
/// assert_eq!(arr_5(), Some(&[1, 10, 100, 1000, 10000]));
///
///
/// const fn err() -> Result<&'static [u64; 5], TryIntoArrayError> {
///     let slice: &[u64] = &[];
///
///     // Letting the function infer the length of the array,
///     try_into_array(slice)
/// }
///
/// assert!(err().is_err());
///
///
/// const fn arr_3() -> &'static [u64; 3] {
///     let slice: &[u64] = &[3, 5, 8];
///
///     let array = result::unwrap!(try_into_array(slice));
///     
///     // You can destructure the array into its elements like this
///     let [a, b, c] = *array;
///     
///     array
/// }
///
/// assert_eq!(arr_3(), &[3, 5, 8]);
///
/// ```
///
#[inline]
pub const fn try_into_array<T, const N: usize>(slice: &[T]) -> Result<&[T; N], TryIntoArrayError> {
    if slice.len() == N {
        let ptr = slice.as_ptr() as *const [T; N];
        unsafe { Ok(&*ptr) }
    } else {
        Err(TryIntoArrayError {
            slice_len: slice.len(),
            array_len: N,
        })
    }
}

/// Tries to convert from `&mut [T]` to `&mut [T; N]`.
///
/// Returns an `Err(TryIntoArrayError{..})` when the slice doesn't match the expected length.
///
/// # Example
///
/// ```rust
/// use konst::{slice, result};
///
/// const fn mut_array_from<const LEN: usize>(slice: &mut [u8], from: usize) -> &mut [u8; LEN] {
///     let sliced = slice::slice_range_mut(slice, from, from + LEN);
///     result::unwrap!(slice::try_into_array_mut(sliced))
/// }
///
/// # fn main() {
///
/// let slice = &mut [3, 5, 8, 13, 21, 34, 55, 89, 144, 233];
///
/// let foo: &mut [u8; 2] = mut_array_from(slice, 0);
/// assert_eq!(foo, &mut [3, 5]);
///
/// let bar: &mut [u8; 3] = mut_array_from(slice, 2);
/// assert_eq!(bar, &mut [8, 13, 21]);
///
/// let baz: &mut [u8; 4] = mut_array_from(slice, 4);
/// assert_eq!(baz, &mut [21, 34, 55, 89]);
///
/// # }
/// ```
///
#[inline]
pub const fn try_into_array_mut<T, const N: usize>(
    slice: &mut [T],
) -> Result<&mut [T; N], TryIntoArrayError> {
    if slice.len() == N {
        unsafe { Ok(&mut *(slice as *mut [T] as *mut [T; N])) }
    } else {
        Err(TryIntoArrayError {
            slice_len: slice.len(),
            array_len: N,
        })
    }
}

#[doc(hidden)]
pub struct __AssertSlice<'a, T> {
    pub x: &'a [T],
}

#[doc(hidden)]
pub struct __AssertSliceMut<'a, T> {
    pub x: &'a mut [T],
}
