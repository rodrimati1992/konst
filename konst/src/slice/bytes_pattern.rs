use crate::{
    chr,
    polymorphism::{HasTypeWitness, MakeTypeWitness, TypeEq, TypeWitnessTypeArg},
};

/// A byte slice pattern.
///
/// Types that implement this trait can be used to search into a byte slice.
///
/// This trait can only be implemented in the `konst` crate.
///
pub trait BytesPattern<const N: usize>: HasTypeWitness<BytesPatternInput<N, Self>> {
    #[doc(hidden)]
    type __Normalized: ?Sized;
}

#[non_exhaustive]
pub enum BytesPatternInput<const N: usize, P: ?Sized + BytesPattern<N>> {
    Str(TypeEq<P, str>, TypeEq<*const P::__Normalized, *const [u8]>),
    Bytes(TypeEq<P, [u8]>, TypeEq<*const P::__Normalized, *const [u8]>),
    Array(
        TypeEq<P, [u8; N]>,
        TypeEq<*const P::__Normalized, *const [u8]>,
    ),
    Char(
        TypeEq<P, char>,
        TypeEq<*const P::__Normalized, *const chr::Utf8Encoded>,
    ),
}

impl<const N: usize, Arg> TypeWitnessTypeArg for BytesPatternInput<N, Arg>
where
    Arg: ?Sized + BytesPattern<N>,
{
    type Arg = Arg;
}

impl MakeTypeWitness for BytesPatternInput<0, str> {
    const MAKE: Self = BytesPatternInput::Str(TypeEq::NEW, TypeEq::NEW);
}
impl BytesPattern<0> for str {
    #[doc(hidden)]
    type __Normalized = [u8];
}

impl MakeTypeWitness for BytesPatternInput<0, [u8]> {
    const MAKE: Self = BytesPatternInput::Bytes(TypeEq::NEW, TypeEq::NEW);
}
impl BytesPattern<0> for [u8] {
    #[doc(hidden)]
    type __Normalized = [u8];
}

impl<const N: usize> MakeTypeWitness for BytesPatternInput<N, [u8; N]> {
    const MAKE: Self = BytesPatternInput::Array(TypeEq::NEW, TypeEq::NEW);
}
impl<const N: usize> BytesPattern<N> for [u8; N] {
    #[doc(hidden)]
    type __Normalized = [u8];
}

impl MakeTypeWitness for BytesPatternInput<0, char> {
    const MAKE: Self = BytesPatternInput::Char(TypeEq::NEW, TypeEq::NEW);
}
impl BytesPattern<0> for char {
    #[doc(hidden)]
    type __Normalized = chr::Utf8Encoded;
}

#[derive(Copy, Clone)]
pub(crate) enum PatternNorm<'a, const N: usize, P>
where
    P: ?Sized + BytesPattern<N>,
{
    Bytes {
        val: &'a [u8],
        ten: TypeEq<*const P::__Normalized, *const [u8]>,
    },
    Char {
        val: chr::Utf8Encoded,
        ten: TypeEq<*const P::__Normalized, *const chr::Utf8Encoded>,
    },
}

impl<'a, const N: usize, P> PatternNorm<'a, N, P>
where
    P: ?Sized + BytesPattern<N>,
{
    pub(crate) const fn new(pattern: &'a P) -> Self {
        match P::WITNESS {
            BytesPatternInput::Str(te, ten) => {
                let val: &'a [u8] = te.in_ref().to_right(pattern).as_bytes();
                PatternNorm::Bytes { val, ten }
            }
            BytesPatternInput::Bytes(te, ten) => {
                let val: &'a [u8] = te.in_ref().to_right(pattern);
                PatternNorm::Bytes { val, ten }
            }
            BytesPatternInput::Array(te, ten) => {
                let val: &'a [u8] = te.in_ref().to_right(pattern);
                PatternNorm::Bytes { val, ten }
            }
            BytesPatternInput::Char(te, ten) => {
                let val: &'a char = te.in_ref().to_right(pattern);
                let val = chr::encode_utf8(*val);
                PatternNorm::Char { val, ten }
            }
        }
    }

    pub(crate) const fn as_bytes(&self) -> &[u8] {
        match self {
            PatternNorm::Bytes { val, ten } => ten.reachability_hint(val),
            PatternNorm::Char { val, ten } => ten.reachability_hint(val.as_bytes()),
        }
    }
}
