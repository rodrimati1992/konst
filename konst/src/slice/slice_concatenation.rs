/// Macro equivalent of `<[&[T]]>::concat`, which takes a constant as an argument.
///
/// This acts like a compile-time-evaluated version of this function:
/// ```rust
/// pub const fn slice_concat<T>(slices: &'static [&'static [T]]) -> [T; LEN]
/// where
///     T: Copy
/// # { [] }
/// # const LEN: usize = 0;
/// ```
///
/// Where `LEN` is the summed length of all inner slices.
///
/// # Example
///
/// ```rust
/// use konst::slice::slice_concat;
///
/// const S: &[&[u8]] = &[&[3, 5], &[8, 13, 21, 34]];
/// assert_eq!(slice_concat!(u8, S), [3, 5, 8, 13, 21, 34]);
///
/// assert_eq!(slice_concat!(u8, &[]), []);
///
/// assert_eq!(slice_concat!(u8, &[&[], &[1, 2, 3], &[4, 5]]), [1, 2, 3, 4, 5]);
///
/// ```
pub use konst_kernel::slice_concat;
