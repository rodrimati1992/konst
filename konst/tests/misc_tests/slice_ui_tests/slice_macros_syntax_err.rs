use konst::slice;

const _: () = {
    let arr = [0u8; 4];

    slice::fill_with!();
    slice::fill_with!(arr);
    slice::fill_with!(arr,);
    slice::fill_with!(arr,| wdwqd);
    slice::fill_with!(arr,||);
    slice::fill_with!(arr,|_|);
};



fn main() {}

