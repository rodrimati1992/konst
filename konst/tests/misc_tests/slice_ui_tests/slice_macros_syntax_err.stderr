error: unexpected end of macro invocation
 --> tests/misc_tests/slice_ui_tests/slice_macros_syntax_err.rs:6:5
  |
6 |     slice::fill_with!();
  |     ^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
  |
note: while trying to match meta-variable `$slice:expr`
 --> src/slice/slice_filler.rs
  |
  |     ($slice:expr, || $v:expr $(,)?) => {
  |      ^^^^^^^^^^^

error: unexpected end of macro invocation
 --> tests/misc_tests/slice_ui_tests/slice_macros_syntax_err.rs:7:26
  |
7 |     slice::fill_with!(arr);
  |                          ^ missing tokens in macro arguments
  |
note: while trying to match `,`
 --> src/slice/slice_filler.rs
  |
  |     ($slice:expr, || $v:expr $(,)?) => {
  |                 ^

error: unexpected end of macro invocation
 --> tests/misc_tests/slice_ui_tests/slice_macros_syntax_err.rs:8:27
  |
8 |     slice::fill_with!(arr,);
  |                           ^ missing tokens in macro arguments
  |
note: while trying to match `||`
 --> src/slice/slice_filler.rs
  |
  |     ($slice:expr, || $v:expr $(,)?) => {
  |                   ^^

error: expected a closure that takes no arguments
 --> tests/misc_tests/slice_ui_tests/slice_macros_syntax_err.rs:9:5
  |
9 |     slice::fill_with!(arr,| wdwqd);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `slice::fill_with` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected expression, found end of macro arguments
  --> tests/misc_tests/slice_ui_tests/slice_macros_syntax_err.rs:10:29
   |
10 |     slice::fill_with!(arr,||);
   |                           --^ expected expression
   |                           |
   |                           while parsing the body of this closure
   |
  ::: src/slice/slice_filler.rs
   |
   |     ($slice:expr, $v:expr $(,)?) => {
   |                   ------- while parsing argument for this `expr` macro fragment
   |
help: you might have meant to open the body of the closure
   |
10 |     slice::fill_with!(arr,|| {);
   |                              +

error: expected a closure that takes no arguments
  --> tests/misc_tests/slice_ui_tests/slice_macros_syntax_err.rs:11:5
   |
11 |     slice::fill_with!(arr,|_|);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `slice::fill_with` (in Nightly builds, run with -Z macro-backtrace for more info)
